src/emscatter.jl

```diff
@@ -0,0 +1,296 @@
+U = TrialFESpace(V,[0])
+
+# # Numerical integration
+# 
+# We generate the triangulation and a second-order Gaussian quadrature for the numerial integration. Note that we create a boundary triangulation from a `Source` tag for the line excitation. Generally, we do not need such additional mesh tags for the source, we can use a delta function to approximate such line source excitation. However, by generating a line mesh, we can increase the accuracy of this source excitation.
+# 
+
+# Generate triangulation and quadrature from model 
+degree = 2
+Ω = Triangulation(model)
+dΩ = Measure(Ω,degree)
+# Source triangulation
+Γ = BoundaryTriangulation(model;tags="Source")
+dΓ = Measure(Γ,degree)
+
+# # PML formulation
+# 
+# More detailed demonstration of the PML formulation can be found in Ref [1]. Here we first define a `s_PML` function: $s(x)=1+\mathrm{i}\sigma(u)/\omega,$ and its derivative `ds_PML`. The parameter `LH` indicates the size of the inner boundary of the PML regions. Finally, we create a function-like object `Λ` that returns the PML factors and define its derivative in gridap. 
+# 
+
+# PML parameters
+R = 1e-4      # Tolerence for PML reflection 
+σ = -3/4*log(R)/d_pml
+LH = [L,H]
+
+# PML coordinate streching functions
+function s_PML(x,σ,k,LH,d_pml)
+    u = abs.(Tuple(x)).-LH./2
+    return @. ifelse(u > 0,  1+(1im*σ/k)*(u/d_pml)^2, $(1.0+0im))
+end
+
+function ds_PML(x,σ,k,LH,d_pml)
+    u = abs.(Tuple(x)).-LH./2
+    ds = @. ifelse(u > 0, (2im*σ/k)*(1/d_pml)^2*u, $(0.0+0im))
+    return ds.*sign.(Tuple(x))
+end
+
+struct Λ<:Function
+    σ::Float64
+    k::Float64
+    LH::Vector{Float64}
+    d_pml::Float64
+end
+
+function (Λf::Λ)(x)
+    s_x,s_y = s_PML(x,Λf.σ,Λf.k,Λf.LH,Λf.d_pml)
+    return VectorValue(1/s_x,1/s_y) 
+end
+
+Fields.∇(Λf::Λ) = x->TensorValue{2,2,ComplexF64}(-(Λf(x)[1])^2*ds_PML(x,Λf.σ,Λf.k,Λf.LH,Λf.d_pml)[1],0,0,-(Λf(x)[2])^2*ds_PML(x,Λf.σ,Λf.k,Λf.LH,Λf.d_pml)[2])
+
+
+# # Weak form
+# 
+# In the mesh file, we labeled the cylinder region with `Cylinder` to distinguish it from other regions. Using this tag, we can assign material properties correspondingly (basically a function with different value in different regions). The weak form is very similar to its mathematical form in gridap.
+# 
+
+labels = get_face_labeling(model)
+dimension = num_cell_dims(model)
+tags = get_face_tag(labels,dimension)
+const cylinder_tag = get_tag_from_name(labels,"Cylinder")
+
+function ξ(tag)
+    if tag == cylinder_tag
+        return 1/ϵ1
+    else
+        return 1.0
+    end
+end
+
+a(u,v) = (∇.*(Λ(σ,k,LH,d_pml)*v))⊙((ξ∘CellField(tags,Ω))*(Λ(σ,k,LH,d_pml).*∇(u))) - (k^2*(v*u))
+
+# Source term (Line source)
+b(v) = ∫(v)*dΓ
+
+
+# # Solver phase
+# 
+# We can assemble the matrix and source term vector with the weak form and then solve for the field.
+# 
+
+A_mat = assemble_matrix(U,V) do u,v
+    ∫( a(u,v) )dΩ
+end
+b_vec = assemble_vector(b_Ω,V)
+u_vec = A_mat\b_vec
+uh = FEFunction(U,u_vec)
+
+# # Analytical solution
+# 
```

Comment:
Add a little introduction:

In this section, we construct the semi-analytical solution to this scattering problem, for comparison to the numerical solution.   This is possible because of the symmetry of the cylinder, which allows us to expand the solutions of the Helmoltz equation in Bessel functions and match boundary conditions at the cylinder interface.  (In 3d, the analogous process with spherical harmonics is known as "Mie scattering".)  For more information on this technique, see [some ref, e.g. Stroud].